File: src/main/java/caches/trigram/TrigramIndex.java
package caches.trigram;

import caches.FileCache;
import caches.Index;
import caches.Revisions;
import caches.changes.*;
import caches.lmdb.LmdbInt2Long;
import caches.records.Revision;
import caches.records.Trigram;
import caches.records.TrigramFile;
import org.lmdbjava.Env;

import java.nio.ByteBuffer;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class TrigramIndex implements Index<TrigramFile, Integer> {

    private final Env<ByteBuffer> env;
    private final TrigramCache cache;
    private final Revisions revisions;
    private final TrigramFileCounterLmdb counter;


    public TrigramIndex(Env<ByteBuffer> env, FileCache fileCache, Revisions revisions) {
        this.env = env;
        cache = new TrigramCache(revisions, new LmdbInt2Long(env, "trigram_pointers"), fileCache);
        this.revisions = revisions;
        counter = new TrigramFileCounterLmdb(this.env, fileCache);
    }

    private static TrigramCounter getTrigramsCount(String str) {
        byte[] bytes = str.getBytes();
        TrigramCounter result = new TrigramCounter();
        for (int i = 2; i < bytes.length; i++) {
            Trigram trigram = new Trigram(new byte[]{bytes[i - 2], bytes[i - 1], bytes[i]});
            result.add(trigram);
        }
        return result;
    }

    @Override
    public void prepare(List<Change> changes) {
        process(changes);
    }

    @Override
    public void processChanges(List<Change> changes) {
        process(changes);
    }

    private void pushActions(TrigramFileCounter deltas, long timestamp) {
        cache.pushCluster(timestamp, deltas);
    }

    @Override
    public Integer getValue(TrigramFile trigramFile, Revision revision) {
        var currentRevision = revisions.getCurrentRevision();
        if (revision.equals(currentRevision)) {
            return counter.get(trigramFile.trigram(), trigramFile.file());
        } else {
            checkout(revision);
            var ans = counter.get(trigramFile.trigram(), trigramFile.file());
            checkout(currentRevision);
            return ans;
        }
    }

    @Override
    public void checkout(Revision targetRevision) {
        try (var txn = env.txnWrite()) {
            var currentRevision = revisions.getCurrentRevision();
            while (!currentRevision.equals(targetRevision)) {
                if (currentRevision.revision() > targetRevision.revision()) {
                    cache.processDataCluster(currentRevision,
                            (bytes, file, d) -> counter.decreaseIt(txn, bytes, file, d));
                    currentRevision = revisions.getParent(currentRevision);
                } else {
                    cache.processDataCluster(targetRevision,
                            (bytes, file, d) -> counter.addIt(txn, bytes, file, d));
                    targetRevision = revisions.getParent(targetRevision);
                }
            }
            txn.commit();
        }
    }


    public void process(List<Change> changes) {
        var delta = new TrigramFileCounter();
        changes.forEach(it -> countChange(it, delta));
        var filteredDelta = new TrigramFileCounter(delta.getAsMap().entrySet().stream()
                .filter(it -> it.getValue() > 0)
                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue)));
        counter.add(filteredDelta);
        if (!changes.isEmpty()) pushActions(delta, changes.get(0).getTimestamp());
    }

    private boolean validateFilename(String filename) {
        return Stream.of(".java"/*, ".txt", ".kt", ".py"*/).anyMatch(filename::endsWith);
    }

    private boolean validateChange(Change change) {
        List<String> filenames = switch (change) {
            case FileChange fileChange -> List.of(fileChange.getPlace().file().getName());
            case FileHolderChange fileHolderChange -> List.of(fileHolderChange.getOldFileName().getName(),
                    fileHolderChange.getNewFileName().getName());
        };
        return filenames.stream().anyMatch(this::validateFilename);
    }

    private void countChange(Change change, TrigramFileCounter delta) {
//            if (!validateChange(change)) return;
        switch (change) {
            case AddChange addChange ->
                    delta.add(addChange.getPlace().file(), getTrigramsCount(addChange.getAddedString()));
            case ModifyChange modifyChange -> {
                delta.decrease(modifyChange.getOldFileName(), getTrigramsCount(modifyChange.getOldFileContent()));
                delta.add(modifyChange.getNewFileName(), getTrigramsCount(modifyChange.getNewFileContent()));
            }
            case CopyChange copyChange ->
                    delta.add(copyChange.getNewFileName(), getTrigramsCount(copyChange.getNewFileContent()));
            case RenameChange renameChange -> {
                delta.decrease(renameChange.getOldFileName(), getTrigramsCount(renameChange.getOldFileContent()));
                delta.add(renameChange.getNewFileName(), getTrigramsCount(renameChange.getNewFileContent()));
            }
            case DeleteChange deleteChange ->
                    delta.add(deleteChange.getPlace().file(), getTrigramsCount(deleteChange.getDeletedString()));
        }
    }

    public TrigramFileCounterLmdb getCounter() {
        return counter;
    }
}

File: src/main/java/com/github/sudu/persistentidecaches/trigram/TrigramIndex.kt
package com.github.sudu.persistentidecaches.trigram

import com.github.sudu.persistentidecaches.Index
import com.github.sudu.persistentidecaches.Revisions
import com.github.sudu.persistentidecaches.changes.*
import com.github.sudu.persistentidecaches.lmdb.CountingCacheImpl
import com.github.sudu.persistentidecaches.lmdb.maps.LmdbInt2Bytes
import com.github.sudu.persistentidecaches.records.Revision
import com.github.sudu.persistentidecaches.records.Trigram
import com.github.sudu.persistentidecaches.records.TrigramFile
import org.lmdbjava.Env
import java.nio.ByteBuffer
import java.nio.file.Path
import java.util.*
import java.util.function.Consumer
import java.util.stream.Stream

class TrigramIndex(
    private val env: Env<ByteBuffer>,
    pathCache: CountingCacheImpl<Path>,
    private val revisions: Revisions
) : Index<TrigramFile, Int> {
    private val cache = TrigramCache(revisions, LmdbInt2Bytes(env, "trigram_deltas"), pathCache)
    @JvmField
    val counter: TrigramFileCounterLmdb = TrigramFileCounterLmdb(this.env, pathCache)
    @JvmField
    val trigramIndexUtils: TrigramIndexUtils = TrigramIndexUtils(this)

    override fun prepare(changes: List<Change>) {
        process(changes)
    }

    override fun processChanges(changes: List<Change>) {
        process(changes)
    }

    private fun pushActions(deltas: TrigramFileCounter, timestamp: Long) {
        cache.pushCluster(timestamp, deltas)
    }

    override fun getValue(trigramFile: TrigramFile, revision: Revision): Int {
        val currentRevision = revisions.currentRevision
        if (revision == currentRevision) {
            return counter.get(trigramFile.trigram, trigramFile.file)
        } else {
            checkout(revision)
            val ans = counter.get(trigramFile.trigram, trigramFile.file)
            checkout(currentRevision)
            return ans
        }
    }

    override fun checkout(targetRevision: Revision) {
        var targetRevision = targetRevision
        var currentRevision = revisions.currentRevision
        env.txnWrite().use { txn ->
//            final var deltasList = new ArrayList<ByteArrIntInt>();
            while (currentRevision != targetRevision) {
                if (currentRevision.revision > targetRevision.revision) {
                    cache.processDataCluster(
                        currentRevision
                    ) { bytes: ByteArray, file: Int, d: Int -> counter.decreaseIt(txn, bytes, file, d) }
                    currentRevision = revisions.getParent(currentRevision)
                } else {
                    cache.processDataCluster(
                        targetRevision
                    ) { bytes: ByteArray, file: Int, d: Int -> counter.addIt(txn, bytes, file, d) }
                    targetRevision = revisions.getParent(targetRevision)
                }
                //                counter.add(txn, deltasList);
//                deltasList.clear();
            }
            txn.commit()
        }
    }


    fun process(changes: List<Change>) {
        val delta = TrigramFileCounter()
        changes.forEach { it: Change -> countChange(it, delta) }

        delta.asMap.entries.stream()
            .filter { it: Map.Entry<TrigramFile?, Int> -> it.value == 0 }
            .map { obj: Map.Entry<TrigramFile?, Int?> -> obj.key }
            .toList()
            .forEach(Consumer { it: TrigramFile? -> delta.asMap.remove(it) })
        counter.add(delta)
        if (!changes.isEmpty()) {
            pushActions(delta, changes[0].timestamp)
        }
    }

    private fun validateFilename(filename: String): Boolean {
        return Stream.of(".java" /*, ".txt", ".kt", ".py"*/).anyMatch { suffix: String? ->
            filename.endsWith(
                suffix!!
            )
        }
    }

    private fun countChange(change: Change, delta: TrigramFileCounter) {
        Objects.requireNonNull(change)
        if (change is AddChange) {
            delta.add(change.place.file, getTrigramsCount(change.addedString))
        } else if (change is ModifyChange) {
            delta.decrease(change.oldFileName, getTrigramsCount(change.oldFileContent))
            delta.add(change.newFileName, getTrigramsCount(change.newFileContent))
        } else if (change is CopyChange) {
            delta.add(change.newFileName, getTrigramsCount(change.newFileContent))
        } else if (change is RenameChange) {
            delta.decrease(change.oldFileName, getTrigramsCount(change.oldFileContent))
            delta.add(change.newFileName, getTrigramsCount(change.newFileContent))
        } else if (change is DeleteChange) {
            delta.decrease(change.place.file, getTrigramsCount(change.deletedString))
        } else {
            throw AssertionError()
        }
    }
    companion object {
        private fun getTrigramsCount(str: String): TrigramCounter {
            val bytes = str.toByteArray()
            val result = TrigramCounter()
            for (i in 2 until bytes.size) {
                val trigram = Trigram(byteArrayOf(bytes[i - 2], bytes[i - 1], bytes[i]))
                result.add(trigram)
            }
            return result
        }
    }
}
